File: lib/rag.ts
/**
 * RAG (Retrieval-Augmented Generation) Pipeline
 *
 * This module handles:
 * 1. Text chunking - splits uploaded documents into smaller segments
 * 2. Embedding generation - creates vector embeddings via OpenAI
 * 3. Vector storage - stores embeddings in an in-memory index
 * 4. Retrieval - finds top-k relevant chunks using cosine similarity
 */

import { embed } from "ai";

// ── Types ──────────────────────────────────────────────────────────────────

export interface DocumentChunk {
  id: string;
  text: string;
  embedding: number[];
}

export interface VectorStore {
  chunks: DocumentChunk[];
}

// ── Global in-memory vector store (per-server instance) ────────────────────

const vectorStores = new Map<string, VectorStore>();

/**
 * Chunk text into smaller overlapping segments.
 * Uses a sliding window approach with configurable size and overlap
 * to maintain context across chunk boundaries.
 */
export function chunkText(
  text: string,
  chunkSize: number = 500,
  overlap: number = 100
): string[] {
  const words = text.split(/\s+/).filter(Boolean);
  const chunks: string[] = [];

  if (words.length <= chunkSize) {
    return [words.join(" ")];
  }

  for (let i = 0; i < words.length; i += chunkSize - overlap) {
    const chunk = words.slice(i, i + chunkSize).join(" ");
    if (chunk.trim().length > 0) {
      chunks.push(chunk);
    }
    if (i + chunkSize >= words.length) break;
  }

  return chunks;
}

/**
 * Generate an embedding vector for a given text using the AI SDK.
 * Uses OpenAI's text-embedding-3-small model via Vercel AI Gateway.
 */
async function generateEmbedding(text: string): Promise<number[]> {
  const { embedding } = await embed({
    model: "openai/text-embedding-3-small" as never,
    value: text,
  });
  return embedding;
}

/**
 * Calculate cosine similarity between two vectors.
 * Returns a value between -1 and 1, where 1 means identical direction.
 */
function cosineSimilarity(a: number[], b: number[]): number {
  let dotProduct = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dotProduct += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  const denominator = Math.sqrt(normA) * Math.sqrt(normB);
  return denominator === 0 ? 0 : dotProduct / denominator;
}

/**
 * Index document text: chunk it, generate embeddings, and store in memory.
 * Returns the session ID to retrieve context later.
 */
export async function indexDocument(
  sessionId: string,
  text: string
): Promise<{ chunksIndexed: number }> {
  const chunks = chunkText(text);

  // Generate embeddings for all chunks in parallel
  const documentChunks: DocumentChunk[] = await Promise.all(
    chunks.map(async (chunkText, index) => {
      const embedding = await generateEmbedding(chunkText);
      return {
        id: `${sessionId}-${index}`,
        text: chunkText,
        embedding,
      };
    })
  );

  // Store in the vector store for this session
  const existing = vectorStores.get(sessionId);
  if (existing) {
    existing.chunks.push(...documentChunks);
  } else {
    vectorStores.set(sessionId, { chunks: documentChunks });
  }

  return { chunksIndexed: documentChunks.length };
}

/**
 * Retrieve the top-k most relevant document chunks for a given query.
 * Uses cosine similarity between the query embedding and stored chunk embeddings.
 */
export async function retrieveContext(
  sessionId: string,
  query: string,
  topK: number = 3
): Promise<string[]> {
  const store = vectorStores.get(sessionId);
  if (!store || store.chunks.length === 0) {
    return [];
  }

  // Generate embedding for the query
  const queryEmbedding = await generateEmbedding(query);

  // Calculate similarity scores for each chunk
  const scored = store.chunks.map((chunk) => ({
    text: chunk.text,
    score: cosineSimilarity(queryEmbedding, chunk.embedding),
  }));

  // Sort by score descending and return top-k
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK).map((s) => s.text);
}

/**
 * Check if a session has any indexed documents.
 */
export function hasDocuments(sessionId: string): boolean {
  const store = vectorStores.get(sessionId);
  return !!store && store.chunks.length > 0;
}

/**
 * Clear all indexed documents for a session.
 */
export function clearSession(sessionId: string): void {
  vectorStores.delete(sessionId);
}

File: rag.py
import numpy as np
from typing import List, Dict, Any, Optional, Union
import os
from openai import OpenAI
import google.generativeai as genai
from dotenv import load_dotenv

load_dotenv()

class DocumentChunk:
    def __init__(self, id: str, text: str, embedding: List[float]):
        self.id = id
        self.text = text
        self.embedding = embedding

class VectorStore:
    def __init__(self):
        self.chunks: List[DocumentChunk] = []

# In-memory vector stores (dictionary by session_id)
vector_stores: Dict[str, VectorStore] = {}

def chunk_text(text: str, chunk_size: int = 500, overlap: int = 100) -> List[str]:
    words = text.split()
    if not words:
        return []
    
    if len(words) <= chunk_size:
        return [" ".join(words)]

    chunks = []
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i : i + chunk_size])
        if chunk.strip():
            chunks.append(chunk)
        if i + chunk_size >= len(words):
            break
    return chunks

def generate_embedding(client: Union[OpenAI, Any], text: str, provider: str = "openai") -> List[float]:
    if provider == "openai":
        response = client.embeddings.create(
            model="text-embedding-3-small",
            input=text
        )
        return response.data[0].embedding
    elif provider == "gemini":
        # client is the genai module or a configured instance
        result = client.embed_content(
            model="models/gemini-embedding-001",
            content=text,
            task_type="retrieval_document"
        )
        return result['embedding']
    else:
        raise ValueError(f"Unsupported provider: {provider}")

def cosine_similarity(a: List[float], b: List[float]) -> float:
    a = np.array(a)
    b = np.array(b)
    dot_product = np.dot(a, b)
    norm_a = np.linalg.norm(a)
    norm_b = np.linalg.norm(b)
    if norm_a == 0 or norm_b == 0:
        return 0
    return dot_product / (norm_a * norm_b)

def index_document(client: Union[OpenAI, Any], session_id: str, text: str, provider: str = "openai") -> int:
    chunks = chunk_text(text)
    
    if session_id not in vector_stores:
        vector_stores[session_id] = VectorStore()
    else:
        vector_stores[session_id] = VectorStore()

    for i, chunk_content in enumerate(chunks):
        embedding = generate_embedding(client, chunk_content, provider)
        chunk = DocumentChunk(
            id=f"{session_id}-{i}",
            text=chunk_content,
            embedding=embedding
        )
        vector_stores[session_id].chunks.append(chunk)
    
    return len(chunks)

def has_documents(session_id: str) -> bool:
    return session_id in vector_stores and len(vector_stores[session_id].chunks) > 0

def retrieve_context(client: Union[OpenAI, Any], session_id: str, query: str, provider: str = "openai", top_k: int = 3) -> List[str]:
    if not has_documents(session_id):
        return []

    # For Gemini, we should ideally use task_type="retrieval_query", 
    # but generate_embedding currently only supports retrieval_document.
    query_embedding = generate_embedding(client, query, provider)
    store = vector_stores[session_id]
    
    scored_chunks = []
    for chunk in store.chunks:
        score = cosine_similarity(query_embedding, chunk.embedding)
        scored_chunks.append((score, chunk.text))
    
    # Sort by score descending
    scored_chunks.sort(key=lambda x: x[0], reverse=True)
    
    return [text for score, text in scored_chunks[:top_k]]
